README.md
# CZ-Parallax jQuery Plugin

Simple and tiny jQuery plugin for Parallax effect.

![](demo.gif)

## Description
This simple plugin is easy to use and with it you can create Parallax effect on any div you want, with as many layers as you want. There are several options (such as speed, zoom level, axis lock) with which you can adjust the plugin to your needs. Also, it is tiny - just **795 bytes** gzipped or **1.59KB** without gzip (minified version).

## Demo
* Please see **demo.html**
* For live demo see [JSFiddle](https://jsfiddle.net/mirzamasic/qx6bmp89/)

## Requirements
* jQuery 1.9.1+

## Installation
* Include necessary JS files

```
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="path-to-file/CZ-Parallax_jQuery-min.js"></script>
```

## Options

### fg
* **Description**: Foreground image link. Only one image link is required.
* **Data type**: string
* **Default value**: none
* **Required**: yes
* **Example value**: https://somewebsite.com/someimage.png

### bgs
* **Description**: Background image links. At least one is required, several can be added.
* **Data type**: array
* **Default value**: none
* **Required**: yes
* **Example value**: ['https://somewebsite.com/someimage.png', 'https://somewebsite.com/someimage2.png', 'https://somewebsite.com/someimage3.png']

### speed
* **Description**: Speed of image movement. Foreground and first background layer images will move in this speed, each additional background layer will move slower than the previous one.
* **Data type**: integer/float
* **Default value**: 1
* **Required**: no
* **Example value**: 1.2 || 3 || .7

### fgZoom
* **Description**: Foreground image zoom. Zoom is useful when using image that covers the div, in order to cover whole div even when moving around. 
* **Data type**: integer/float
* **Default value**: 1
* **Required**: no
* **Example value**: 1.2 || 3 || .7

### bgZoom
* **Description**: Background layers image zoom. Zoom is useful when using image that covers the div, in order to cover whole div even when moving around. 
* **Data type**: integer/float
* **Default value**: 1.05
* **Required**: no
* **Example value**: 1.2 || 3 || .7

### lock
* **Description**: Locking movement to only one axis. By default, it is unlocked - movement is available to both X and Y axis.
* **Data type**: string
* **Default value**: none
* **Required**: no
* **Example value**: 'x' || 'X' || 'y' || 'Y'

## Usage
Example code:

```
    // Minimum setup with only required options
    $('#some-element').CZParallax({
        fg: 'https://somewebsite.com/foreground-image.png',
        bgs: ['https://somewebsite.com/background-image.png']
    });

    // Setup with all options set
    $('#some-element').CZParallax({
        fg: 'https://somewebsite.com/foreground-image.png',
        bgs: [
            'https://somewebsite.com/background-image.png',
            'https://somewebsite.com/background-image2.png',
            'https://somewebsite.com/background-image3.png'
        ],
        speed: 2.5,
        fgZoom: 1.1,
        bgZoom: 1.5,
        lock: 'x'
    });
```

## License

The expandable plugin is licensed under the MIT License.

Copyright (c) 2019 Mirza Mašić


CHANGELOG.md
# CZ-Parallax jQuery Plugin CHANGELOG

## 1.0.0
* First stable release

js/CZ-Parallax_jQuery-min.js
(function(c){c.fn.CZParallax=function(k){var a=c.extend({speed:1,bgZoom:1.05,fgZoom:1,lock:"",fg:"",bgs:[]},k);this.each(function(){var b=c(this),e=b.width(),f=b.height();if(Number(a.speed)!==a.speed||Number(a.bgZoom)!==a.bgZoom||Number(a.fgZoom)!==a.fgZoom)throw"CZP ERROR: Parameters speed, bgZoom and fgZoom need to be numbers.";0>=a.speed&&(a.speed=1);if(""===a.fg)throw"CZP ERROR: Foreground image link not supplied.";b.addClass("cz cz-foreground");b.wrap('<div style="overflow:hidden;width:'+e+"px;height:"+
f+'px;""></div>');b.css({"background-image":"url("+a.fg+")",transform:"scale("+a.fgZoom/a.bgZoom+")"});var g=a.bgs.length;if(1>g)throw"CZP ERROR: No background links supplied. Minimum 1 link is required.";a.bgs.map(function(c,d){b.wrap('<div class="cz cz-background-'+(d+1)+'" style="background-image:url('+c+");height:"+f+"px;width:"+e+"px;transform:scale("+(0===d?a.bgZoom:1)+')"></div>')});b.mousemove(function(b){var d=(e/2-b.pageX)/40*a.speed,h=(f/2-b.pageY)/40*a.speed;switch(a.lock){case "x":case "X":d=
0;break;case "y":case "Y":h=0;break;case "":case "none":break;default:throw'CZP ERROR: Parameter lock can only be "x", "X", "y", "Y" or empty (not included).';}c(".cz-foreground").css({"background-position":"calc(50% - "+d+"px) calc(50% - "+h+"px)","background-size":"90% auto"});a.bgs.map(function(a,b){c(".cz-background-"+(b+1)).css("background-position","calc(50% + "+d/(g-b)+"px) calc(50% + "+h/(g-b)+"px)")})})});c(".cz").css({"background-repeat":"no-repeat","background-position":"center","background-size":"cover"});
c(".cz-foreground").css("background-size",90*a.fgZoom+"% auto")}})(jQuery);

js/CZ-Parallax_jQuery.js
(function ($) {
    $.fn.CZParallax = function(options) {
        // Default settings
        let settings = $.extend({
        speed: 1,
        bgZoom: 1.05,
        fgZoom: 1,
        lock: '',
        fg: '',
        bgs: []
        }, options );
        
        this.each(function() {
            let _this = $(this);
            let W = _this.width();
            let H = _this.height();
            
            // Checking if parameters provided are numbers
            if(Number(settings.speed) !== settings.speed || Number(settings.bgZoom) !== settings.bgZoom || Number(settings.fgZoom) !== settings.fgZoom) throw 'CZP ERROR: Parameters speed, bgZoom and fgZoom need to be numbers.';

            // If speed parameter provided is 0 or less than 0, set it to 1
            if(settings.speed <= 0) settings.speed = 1;

            // Foreground image has to be provided
            if(settings.fg === '') {
                throw 'CZP ERROR: Foreground image link not supplied.';
            }
            else {
                _this.addClass('cz cz-foreground');
                _this.wrap('<div style="overflow:hidden;width:'+W+'px;height:'+H+'px;""></div>');
                _this.css({'background-image': 'url('+settings.fg+')', 'transform': 'scale('+settings.fgZoom/settings.bgZoom+')'});
                
                let L = settings.bgs.length;
                // At least one background image has to be provided
                if(L < 1) {
                    throw 'CZP ERROR: No background links supplied. Minimum 1 link is required.';
                }
                else {
                    settings.bgs.map((v, i) => {
                        let Z = i === 0 ? settings.bgZoom : 1;
                        _this.wrap('<div class="cz cz-background-'+(i+1)+'" style="background-image:url('+v+');height:'+H+'px;width:'+W+'px;transform:scale('+Z+')"></div>');
                    });

                    _this.mousemove(e => {
                        let X = e.pageX;
                        let Y = e.pageY;
                        let bgX = ((W/2)-X)/40*settings.speed;
                        let bgY = ((H/2)-Y)/40*settings.speed;
                        
                        // Direction locking
                        switch(settings.lock){
                            case 'x':
                            case 'X':
                                bgX = 0;
                                break;
                            case 'y':
                            case 'Y':
                                bgY = 0 
                                break;
                            case '':
                            case 'none':
                                break;
                            default:
                                throw 'CZP ERROR: Parameter lock can only be "x", "X", "y", "Y" or empty (not included).'
                        }
                        
                        // Foreground image movement
                        $(".cz-foreground").css({'background-position': 'calc(50% - '+bgX+'px) calc(50% - '+bgY+'px)', 'background-size': '90% auto'});

                        // Background image movement
                        settings.bgs.map((v, i) => {
                            $('.cz-background-'+(i+1)).css('background-position', 'calc(50% + '+bgX/(L-i)+'px) calc(50% + '+bgY/(L-i)+'px)');
                        });              
                    });
                }
            }
        });
        $('.cz').css({
            'background-repeat': 'no-repeat',
            'background-position': 'center',
            'background-size': 'cover'
        });
        $('.cz-foreground').css('background-size', 90*settings.fgZoom+'% auto');
    };
}( jQuery ));